_bleio: support anonymous advertising

xobs opened the pull request:
This experimental patch adds a new parameter to `_bleio.adapter.start_advertising()`: `anonymous`.

This boolean parameter defaults to `False`. When set to `True`, it enables privacy mode by calling `sd_ble_gap_privacy_set()`.

I have tested this on my device, and I confirm that it works -- the address is randomized, and is rotated every 15 minutes.  A second device is capable of connecting to a GATT service and communicating.

Unanswered questions:

- How does this work with bonding?
- Should we expose the cycle time? It's currently set to `0` which uses the Nordic defaults
- Should the `IRK` be exposed anywhere, and does this affect bonding at all?

tannewt commented on shared-bindings/_bleio/Adapter.c:
```diff
@@ -145,17 +145,19 @@ const mp_obj_property_t bleio_adapter_name_obj = {
 //|         :param buf data: advertising data packet bytes
 //|         :param buf scan_response: scan response data packet bytes. ``None`` if no scan response is needed.
 //|         :param bool connectable:  If `True` then other devices are allowed to connect to this peripheral.
+//|         :param bool anonymous:  If `True` then this device's MAC address is randomized regularly.
```
We should stop advertising when the address rotates. Keeping the advertising data static while rotating the mac address removes the anonymization aspect of it.

xobs commented:
I added a new property `timeout` that lets a user specify a number of seconds for the advertisement to time out. If no timeout is specified when anonymous mode is requested, then the maximum permitted timeout is selected (currently 180 seconds).

Additionally, I created a new property `.advertising` that indicates whether the device is still advertising.

dhalbert commented on ports/nrf/common-hal/_bleio/Adapter.c:
```diff
@@ -594,7 +595,26 @@ STATIC void check_data_fit(size_t data_len, bool connectable) {
+
+    switch (ble_evt->header.evt_id) {
+        case BLE_GAP_EVT_ADV_SET_TERMINATED:
+            mp_printf(&mp_plat_print, "Advertising set terminated - %d advertising events were completed - reason: %d\n", ble_evt->evt.gap_evt.params.adv_set_terminated.num_completed_adv_events, ble_evt->evt.gap_evt.params.adv_set_terminated.reason);
```
This looks like a debugging printout you meant to remove, comment out, or guard with an `#if` or `#ifdef`.

xobs replied:
Good catch. I removed that.

dhalbert commented on ports/nrf/common-hal/_bleio/Adapter.c:
```diff
@@ -594,7 +595,26 @@ STATIC void check_data_fit(size_t data_len, bool connectable) {
+
+        default:
+            // For debugging.
+            mp_printf(&mp_plat_print, "Unhandled Advertising etvent: 0x%04x\n", ble_evt->header.evt_id);
```
Also debugging (and it says "etvent" instead of "event")

xobs replied:
I corrected the spelling on this one and commented it out, like the others.

dhalbert commented on shared-bindings/_bleio/Adapter.c:
```diff
@@ -182,11 +189,13 @@ STATIC mp_obj_t bleio_adapter_start_advertising(mp_uint_t n_args, const mp_obj_t
     }
 
     bool connectable = args[ARG_connectable].u_bool;
+    bool anonymous = args[ARG_anonymous].u_bool;
+    uint32_t timeout = args[ARG_timeout].u_int;
```
Since the common_hal routine will ignore `anonymous` if there's no timeout, could you add a validation check for this, and complain if `anonymous` is set without a timeout?

xobs replied:
Okay, if you specify anonymous, it now requires you specify a timeout, rather than picking one.

dhalbert replied:
Sorry, I misunderstood the `common_hal` routine. I thought it was ignoring the `anonymous` if there as no timeout, but that was just because I was missing some code due to the way the diff displayed. It's fine to not specify the timeout if `anonymous` is `True`, as long as the constructor's documentation says that, and says what the timeout will be. So I believe your old logic is OK, if it's as stated above; just add more doc to the `timeout` arg description.

dhalbert commented on ports/nrf/common-hal/_bleio/Adapter.c:
```diff
@@ -669,6 +715,27 @@ void common_hal_bleio_adapter_start_advertising(bleio_adapter_obj_t *self, bool
+    // point of randomizing the MAC address?
+    if (!timeout) {
+        if (anonymous) {
+            // The Nordic macro is in units of 10ms. Convert to seconds.
+            uint32_t adv_timeout_max_secs = BLE_GAP_ADV_TIMEOUT_LIMITED_MAX / 100;
```
You could add a `UNITS_TO_SEC` macro to `ble_drv.h` to do the reverse of `SEC_TO_UNITS`.

xobs commented:
Thanks for the feedback, @dhalbert . I've made all of those changes as requested.

xobs commented:
I pushed to the wrong repo. Oops.

Also, alright, I reverted that change. I already added a note to the documentation regarding defaults of `timeout` when `anonymous` was specified.
